### PR TITLE
Issue/6836 enable back dating drafts

### LINKED ISSUE #6836: Can't set a date in the past for draft/new posts
A user reported in HS that cannot set the publish date for a new post in the past.
637537e123ea97530973c1e82104630c91ee59b7

On already published posts the user can set a "back-date" without issues.

### PR DESCRIPTION
**Pull request by mzorz**
Fixes #6836 

This PR aims at bringing the Posts's date handling as closest to par as possible with wp-admin/Calypso.

*Note:* we can change the labels an as we wish, I only copied them in our strings here so it's easier to follow the code as per the wp-admin reference.

To test:
Using Classic/Gutenberg editors should be the same for these tests.

Update release notes:
CASE A: draft, no saves
1. start a draft
2.  do not enter anything (no title, no body)
3. tap on overflow menu icon and go to Post Settings
4. Observe it reads "Publish: Immediately"

CASE B.1: draft, enter data
1. start a draft
2. enter a title and/or something in the body
3. tap on overflow menu icon and go to Post Settings
4. Observe it reads "Publish: Immediately"

CASE B.2: draft, save
1. start a draft
2. enter a title and/or something in the body
3. tap back so the draft saves
4. open the draft in the Editor again
5. tap on overflow menu icon and go to Post Settings
6. Observe it reads ~"Publish: Immediately"~ "Publish on: <date>"

CASE C: draft, change publish date in the future (do not tap Schedule)
1. start a draft
1.b (test both entering a title and text and also test leaving everything empty)
2. tap on the overflow menu icon and go to Post settings
3. change publish date in the future (i.e. tomorrow)
4. observe it changes to show the date as expected, and the button on the action bar changes to `SCHEDULE`

CASE D: draft, change publish date in the past, publish
1. start a draft
1.b (test both entering a title and text and also test leaving everything empty)
2. tap on the overflow menu icon and go to Post settings
3. change publish date in the past (i.e. yesterday)
4. observe it shows "Publish immediately" (as the draft has not been saved);  the button on the action bar reads `PUBLISH`.
5. tapping PUBLISH will trigger the publish dialog (promo dialog first time,  or confirmation dialog other times), and if you continue publishing it will be published with the chosen date (note the Posts list will be sorted by date, so, find your Post there)

~CASE E: draft, change publish date in the past, save draft
Repeat steps 1-4 of above case D.~

~6. tapping BACK instead, or tapping "Save as draft" in the menu will save the back-dated draft (will show as "Draft" in the Posts list, look for it among the posts of previous days).
7. verify going to Posts settings will show Publish "Immediately" and main action  "UPDATE".
8. verify opening the overflow menu shows "Publish now".
9. If you tap UPDATE, note it will save your changes but will be kept as draft.
10. if you tap "Publish now "  on the overflow menu icon, it will show the publish confirmation dialog (and publish if you confirm).~

CASE E: draft, change publish date in the past, save draft
Repeat steps 1-4 of above case D.

6. tapping BACK instead, or tapping "Save as draft" in the menu will save the back-dated draft (will show as "Draft" in the Posts list, look for it among the posts of previous days).

CASE F: edit back-dated draft
1. open the recently saved Post in CASE E.
2. verify going to Posts settings will show "Publish on <date>" and main action  "UPDATE".
3. verify opening the overflow menu shows "Publish now".
4. If you tap UPDATE, note it will save your changes but will be kept as draft.
5. if you tap "Publish now "  on the overflow menu icon, it will show the publish confirmation dialog (and publish if you confirm).

- [X] If there are user facing changes, I have added an item to `RELEASE-NOTES.txt`.

### CODE REVIEW

**Comment by mzorz** [comment]
Note for the reviewers: I'm targeting the `release/12.0` branch given the "draft" default has been found to accentuate the problem in the current alpha, but we can target `develop` for more testing,  or even plan a hotfix otherwise as we see usage during the following weeks. cc @jtreanor

**Comment by daniloercoli** [comment]
> CASE B.2: draft, save

>start a draft
enter a title and/or something in the body
tap back so the drafts saves
open the draft in the Editor agai
tap on overflow menu icon and go to Post Settings
Observe it reads "Publish: Immediately"


When I re-opened the Post settings screen it does't say `Publish: Immediately`., but rather there is the date set when the draft was saved on the server.

**Comment by mzorz** [comment]
> When I re-opened the Post settings screen it does't say Publish: Immediately., but rather there is the date set when the draft was saved on the server.

Ah, that is ok, I forgot to update the test case. Yes labels will change once the server gets updated, only case where "immediately" appears should be while it hasn't been set a date and saved to the server yet

**Comment by daniloercoli** [comment]
> CASE D: draft, change publish date in the past, publish

I made a variation of this test case, by trying to publish in the past a post that was previously scheduled for tomorrow.

When I changed the date back to past, I see both the label set as `Scheduled for: 22 Mar...` , and the top right button is `Schedule` (Pianifica).
device-2019-03-25-110858

**Comment by mzorz** [comment]
hi @daniloercoli 👋 - just edited the test cases as CASE E was wrong (copy paste mistake on my part), and split it into CASE F as well for clarity. Also edited CASE B.2 as read above.

**Comment by mzorz** [comment]
>  a variation of this test case, by trying to publish in the past a post that was previously scheduled for tomorrow.

Fixed in bbc4d5e41e. Description:
- setting to scheduled (i.e. set publish date to tomorrow) and tapping back will show SCHEDULED, then set it back to yesterday, shows PUBLISHED ON: and it should again show Immediately (as it should still be a local draft). This is fixed by bbc4d5e41e. Tapping on it will show the date/time picker set to the actual value.

- setting to `scheduled`, then SAVE (tap BACK) will save a `local draft` (instead of sending the draft to the server, which is OK as otherwise it would be saved as `SCHEDULED` on the server), then open it again, should show SCHEDULED (and label "Scheduled for: _date_"). Setting it back to YESTERDAY should change it to DRAFT (show IMMEDIATELY). Tapping BACK now does save the copy to the server as it's a DRAFT.

Ready for another round @daniloercoli 🙇

**Comment by mzorz** [comment]
Reported this case elsewhere:

1. start a draft,
2. schedule it (set date in the future) and tap on `SCHEDULE` button  (this makes it upload),
3. _then_ open it and edit again
4. now try changing the date back to the past --> observe the label is still SCHEDULE

Added case to address this in 5255d69e2b:  when a user back-dates an already SCHEDULED post: convert it back to DRAFT as we don't want it to get published by mistake.

**Comment by daniloercoli** [comment]
it seems to work as expected, may worth another set of 👁 to give it a spin, since it's a sensitive part of the app.

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostSettingsFragment.java [hunk 1]
```diff
@@ -723,16 +739,32 @@ private void updatePublishDateTextView() {
...
-        if (PostUtils.shouldPublishImmediately(postModel)) {
-            mPublishDateTextView.setText(R.string.immediately);
-        } else {
-            String dateCreated = postModel.getDateCreated();
-            if (!TextUtils.isEmpty(dateCreated)) {
-                String formattedDate = DateUtils.formatDateTime(getActivity(),
-                                                                DateTimeUtils.timestampFromIso8601Millis(dateCreated),
-                                                                getDateTimeFlags());
-                mPublishDateTextView.setText(formattedDate);
+        String labelToUse = "%s"; // default label to use is just the formatted date
```
Comment (oguzkocer) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostSettingsFragment.java]: Did you maybe forgot to remove or change the `"%s"` since that's not a valid string by itself? I think we can simply remove the initialization as it should be initialized later unless we have a good default value.

**Comment by daniloercoli** [comment]
> CASE D: draft, change publish date in the past, publish > I am just curious, shouldn't the post settings say "Publish on: {Selected past date}" instead of immediately? It feels pretty weird to take an action as a user only to see that nothing has changed in the UI. 

Yes, we've had a long discussion in Slack about this behavior, and decided to open a new issue for it asking for design help. There are some edge cases we might want to consider to be in par with the web and input from design will clear our doubts.
Also consider that `"Publish on: immediately"`  it's the current behavior of `develop`, for already published post, and was there since long time. 

Side note - Why we need some design help: On mobile there is also the ability of tapping Back arrow on the top toolbar, option that is missing on the web. In the case the user taps Back, we should decide what to do, since tapping back again in the editor "main" screen will update the post on the server. Maybe leading on undesired changes sent to the remote server without confirmation.

File: WordPress/src/main/java/org/wordpress/android/ui/posts/PostUtils.java [hunk 2]
```diff
@@ -281,6 +281,13 @@ static boolean shouldPublishImmediately(PostModel postModel) {
         return pubDate == null || !pubDate.after(now);
     }
 
+    static boolean isPublishDateInTheFuture(PostModel postModel) {
+        Date pubDate = DateTimeUtils.dateFromIso8601(postModel.getDateCreated());
+        Date now = new Date();
+        // Publish immediately for posts that don't have any date set yet and drafts with publish dates in the past
```
[the remark this answers is not in the document]
Comment (mzorz) [reply WordPress/src/main/java/org/wordpress/android/ui/posts/PostUtils.java]: oops! that was a leftover from when the method was added. Removed in 8aaccbf9c8

**Comment by mzorz** [comment]
Hi @oguzkocer thank you very much for the review! I found the comments useful and the suggestions certainly make the code better so, applied your patch in 2eccae5bcf 🙇 .

As for this:

> CASE B.2: draft, save -> After following the steps, if I tap on the date and select immediately, nothing happens.

Tested it - it does change it (tapping `Immediately` on the date picker then tapping back twice will make the Post upload, that's because it effectively has changes). The thing is, you won't see any apparent change if you tap Immediately at the same minute that was already set 😃 .
Basically, perform the same steps and wait til the minute changes and you'll see the change.

I think this should be good to go! And we'll open a new issue for the discussion about `Immediately` functionality as @daniloercoli  explained in https://github.com/wordpress-mobile/WordPress-Android/pull/9466#issuecomment-476514270

**Comment by mzorz** [comment]
Merging as agreed to create a new beta soon 👍 
cc @jtreanor
